web-console/src/lib/compositions/streaming/management/usePipelineMetrics.ts

```diff
@@ -5,24 +5,47 @@ import { tuple } from '$lib/functions/common/tuple'
 import {
   ControllerStatus,
   GlobalMetrics,
+  GlobalMetricsTimestamp,
   InputEndpointMetrics,
   OutputEndpointMetrics,
   PipelineStatus
 } from '$lib/types/pipeline'
 import { useState } from 'react'
+import invariant from 'tiny-invariant'
 
 import { useQuery } from '@tanstack/react-query'
 
-const toMetrics = (data: ControllerStatus, { timeMs }: { timeMs: number }) => ({
-  input: new Map(data.inputs.map(cs => tuple(cs.config.stream, cs.metrics))),
-  output: new Map(data.outputs.map(cs => tuple(cs.config.stream, cs.metrics))),
-  global: [{ ...data.global_metrics, timeMs }]
-})
-
 const emptyData = {
   input: new Map<string, InputEndpointMetrics>(),
   output: new Map<string, OutputEndpointMetrics>(),
-  global: []
+  global: [] as GlobalMetricsTimestamp[]
+}
+
+type PipelineMetrics = typeof emptyData
+
+/**
+ * Accumulate metrics history, accounting for desired time window, dropping previous data if the new data overwrites it
+ * @param keepMs Time window of metrics age to be kept
+ */
+const reconcileHistoricData = (
```

Comment:
I think this is much better.
The comments help a lot too, as does the function name